set_fm_data.py

```diff
@@ -0,0 +1,475 @@
+from datetime import datetime, timedelta
+import time
+import json
+import math
+import re
+import requests
+from typing import AsyncGenerator, List, Optional
+import appdaemon.plugins.hass.hassapi as hass
+
+from wallbox_client import WallboxModbusMixin
+
+
+# ToDo:
+# Onder de 20% ook niet availabel..
+# Start tijden van post data kloppen niet.
+
+class SetFMdata(hass.Hass, WallboxModbusMixin):
+    # I account and send results to FM hourly for intervals @ resolution, eg. 1/12th of an hour:
+    # + Charged power in kWh
+    # + Availability of car and charger for automatic charging (% of time)
+
+    # We receive power changes at irregular intervals. Usuallly about 15 seconds apart but sometimes hours.
+    # We want to create power usages in regular intervals of eg. 1/12th of an hour.
+    # Power changes: |    |    |  |    |     |               | (called periods)
+    # Intervals:        |        |        |        |        |        |
+
+    # Access token for FM
+    fm_token: str
+
+    # At what time intervals is the chargepower stored (minutes)
+    readings_resolution: int
+    hourly_power_readings_since: datetime
+    hourly_availability_readings_since: datetime
+    hourly_soc_readings_since: datetime
+
+    # Variables to help calculate avarage power over the last readings_resolution minutes
+    current_power_since: datetime
+    current_power: int
+    # Duration between two changes in power
+    power_period_duration: float
+    period_power_x_duration: float
+    # Holds the weighted_avagared power readings of the last hour untill sent to backend.
```

Comment:
Weighted because it is power*duration

Reply:
After much thought, I get your rationale now, but the variable names are still confusing to me, and I think that is also because they are used inconsistently.

In the main docstring you mention three intervals, one of them implicitly:
- "irregular intervals", which you call "periods" (even though they aren't periodic)
- "regular intervals", which you also call "readings resolution" (and also just intervals)
- "sends results to FM hourly", which also implies a regular interval
```
App receives power changes at irregular intervals. Usually about 15 seconds apart but sometimes hours.
We want to create power usages in regular intervals of e.g. 1/12th of an hour.
Power changes: |    |    |  |    |     |               | (called periods)
Intervals:        |        |        |        |        |        |
```
I would recommend more consistent terminology and making clear definitions early on:
- Power changes occur at irregular intervals (readings): usually about 15 seconds apart but sometimes hours
- We derive a time series of power with a regular interval (that is, with a fixed period): we chose 5 minutes
- We send the time series to FlexMeasures in batches, periodically: we chose once every 1 hour

In lines 192-193 it looks to me like you are essentially redefining what a period is, because its duration is slowly growing.
```
self.period_power_x_duration += (duration * power)
self.power_period_duration += duration
```
I would describe this differently: you are tracking the energy accumulated within the regular interval of the time series of power. In other words, it shouldn't be an attribute of your "irregular period", but an attribute of your "regular interval". Maybe something like:
```
energy_since_last_power_change = duration * power
self.accumulated_energy += energy_since_last_power_change
```

In lines 209-2010 you are finalizing your weighted average calculation:
```
average_period_power = round((self.period_power_x_duration / self.power_period_duration)/1000000, 5)
self.power_readings.append(average_period_power)
```
When I'm reading this, I'm thinking: `self.period_power_x_duration` is a quantity of energy accumulated within some period (in Ws, i.e. watt-seconds, or equivalently, Joule), and you divide it over the duration of that period, in seconds (`self.power_period_duration`), which gives you the average power during that period.

But this is not the "irregular period" you defined in the docstring anymore, right? By now its duration has grown to become the duration of the "regular interval", which is `readings_resolution`.

Reply:
Moved to https://github.com/SeitaBV/v2g-liberty/issues/28.